server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ws/ReactiveWebSocketHandler.scala

```diff
@@ -0,0 +1,213 @@
+      val res = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.INTERNAL_SERVER_ERROR)
+      res.headers().set(HttpHeaderNames.CONTENT_LENGTH, 0)
+      res.headers.set(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE)
+      val _ = ctx.writeAndFlush(res).close()
+    }
+
+    def replyWith503(): Unit = {
+      val res = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.SERVICE_UNAVAILABLE)
+      res.headers().set(HttpHeaderNames.CONTENT_LENGTH, 0)
+      res.headers.set(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE)
+      val _ = ctx.writeAndFlush(res).close()
+    }
+
+    def replyNotFound(): Unit = {
+      val res = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.NOT_FOUND)
+      res.headers.set(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE)
+      val _ = ctx.writeAndFlush(res).close()
+    }
+
+    def replyWithRouteResponse(
+        req: HttpRequest,
+        serverResponse: ServerResponse[NettyResponse],
+        channelPromise: ChannelPromise,
+        byteBuf: ByteBuf
+    ): Unit = {
+      val res = new DefaultFullHttpResponse(req.protocolVersion(), HttpResponseStatus.valueOf(serverResponse.code.code), byteBuf)
+      res.setHeadersFrom(serverResponse, serverHeader)
+      res.headers().set(HttpHeaderNames.CONTENT_LENGTH, byteBuf.readableBytes())
+      val _ = ctx.writeAndFlush(res, channelPromise)
+    }
+
+    msg match {
+      case req: FullHttpRequest if isWsHandshake(req) =>
+        if (isShuttingDown.get()) {
+          logger.info("Rejecting WS handshake request because the server is shutting down.")
+          replyWith503()
+        } else {
+          ReferenceCountUtil.release(msg)
+          val (runningFuture, _) = unsafeRunAsync { () =>
+            route(NettyServerRequest(req.retain()))
+              .map {
+                case Some(response) => response
+                case None           => ServerResponse.notFound
+              }
+          }
+
+          runningFuture.onComplete {
+            case Success(serverResponse) if serverResponse == ServerResponse.notFound =>
+              replyNotFound()
+            case Success(serverResponse) =>
+              try {
+                serverResponse.body match {
+                  case Some(function) => {
+                    val content = function(ctx)
+                    content match {
+                      case r: ReactiveWebSocketProcessorNettyResponseContent =>
+                        initWsPipeline(ctx, r, req)
+                      case ByteBufNettyResponseContent(channelPromise, byteBuf) =>
+                        // Handshake didn't return a Pipe, but a regular response. Returning it back.
+                        replyWithRouteResponse(req, serverResponse, channelPromise, byteBuf)
+                      case otherContent =>
+                        // An unsupported type of regular response, returning 500
+                        replyWithError500(
```

Comment:
one problem here: if a regular endpoint is hit with WS headers, it might end up with a 500 and an error log - so I'm not sure if we shouldn't run regular processing in this case

Reply:
We do run regular processing in line 133 above, if the interpreter returns a standard, non-streaming, non-file response.

Reply:
Yes, but there's still an easily achievable situation where a client error (calling any endpoint - including streaming/file ones) with the incorrect headers results in a server error

maybe this can even be considered a security problem? as you can easily spam a server with `logger.error` (which might trigger monitoring)

Reply:
Ah ok, I get it now. How about changing this 500 to a 400 without logging then?

Reply:
Ah yes sounds good. As long as we don't leak any resources? (connection counters, open files, open streams?)

Reply:
Actually an open file or a stream (Publisher[]) might get leaked here if not handled. I'll take that into account, thanks!